spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+before proceeding, it will deadlock. Executing `broadcast_tx` calls
+involves acquiring the mempool lock that Tendermint holds during the `Commit` call.
+If you make the call to the `broadcast_tx` endpoints concurrently,
+that's not a problem, it just can't be part of the sequential logic of the
+`Commit` function.
+
+#### Candidate States
+
+Tendermint calls `PrepareProposal` when it is about to send a proposed block to the network.
+Likewise, Tendermint calls `ProcessProposal` upon reception of a proposed block from the
+network. In both cases, the proposed block's data (last block's commit info, transactions, etc.)
+is disclosed to the Application, in the same conditions as is done in `FinalizeBlock`.
+
+The Application may decide to *immediately* execute the given block (i.e., upon `PrepareProposal`
+or `ProcessProposal`). There are two main reasons why the Application may want to do this:
+
+* In order to be sure that the block does not contain *any* invalid transaction, there is usually
+  no way other than fully executing the transactions in the block as though it was the *decided*
+  block.
+* Upon reception of the decided block via `FinalizeBlock`, if that same block was executed
+  upon `PrepareProposal` or `ProcessProposal` and the resulting state was kept in memory, the
+  Application can simply apply that state (faster) to the main state, rather than reexecuting
+  the decided block (slower).
+
+`PrepareProposal`/`ProcessProposal` can be called many times for a given height. Moreover,
+it is not possible to accurately predict which of the blocks proposed in a height will be decided, being delivered to the application
+in that height's `FinalizeBlock`.
+Therefore, the state resulting from executing a proposed block, denoted a *candidate state*, should
+be kept in memory as a possible final state for that height. When `FinalizeBlock` is called, the Application should
+check if the decided block corresponds to one of its candidate states; if so, it will apply it as
+its `ExecuteTxState` (see [Consensus Connection](#consensus-connection) below),
+which will be persisted during the upcoming `Commit` call.
+
+In bad runs (e.g., under network instability), Consensus might take many rounds.
+In this case, several proposed blocks will be disclosed to the Application for a given height.
+Indeed, by the nature of Consensus, the number of proposed blocks received by the Application for
+a particular height cannot be bound, so Application developers must act with care and use mechanisms
+to bound memory usage. As a general rule, the Application should be ready to discard candidate states
+before `FinalizeBlock`, even if one of them might end up corresponding to the
+decided block and thus have to be reexecuted upon `FinalizeBlock`.
+
+### States and ABCI++ Connections
+
+#### Consensus Connection
+
+The Consensus Connection should maintain an `ExecuteTxState` - the working state
```

Comment:
The typography `ExecuteTxState` seems to imply this is a data structure defined by Tendermint, but I think this is an application implementation detail. I think the key point is something like:

> The application must track which candidate execution state will be reported as the latest committed state during the `Commit` operation on the consensus connection. This value is updated by the call to `FinalizeBlock`, which identifies the decided block. Block executions computed during `PrepareProposal` or `ProcessProposal` **must not** update this state.

or words to that effect. Or maybe just call out specifically that `ExecuteTxState` is not a specific data structure that we're identifying, it's a notional thing we're asserting the application has to keep track of in whatever way makes sense.

Reply:
I'll keep the terms ("ExecuteTxState", "CheckTxState", etc) because they're referenced at several places in text later on. So I don't want to get into coming up with a complete new way of explaining all this.
Having said that, I agree with you that the typography used is confusing in that it matches the one used for _actual_ data structures or method names. I'll change it to italic. I think that'll eliminate the confusion.